# Arcadia-Science/noveltree: Citations

## [nf-core](https://doi.org/10.1038/s41587-020-0439-x)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. (2020). The nf-core framework for community-curated bioinformatics pipelines. https://doi.org/10.1038/s41587-020-0439-x

## [Nextflow](https://doi.org/10.1038/nbt.3820)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. (2017) Nextflow enables reproducible computational workflows. https://doi.org/10.1038/nbt.3820

## Pipeline tools - Software and Packages

- [BUSCO](https://doi.org/10.1002/cpz1.323)

  > Manni M, Berkeley MR, Seppey M, Zdobnov EM. (2021). BUSCO: Assessing genomic data quality and beyond. https://doi.org/10.1002/cpz1.323

- [UniProt](https://doi.org/10.1093/nar/gkac1052)

  > The UniProt Consortium. (2023). UniProt: the universal protein knowledgebase in 2023. https://doi.org/10.1093/nar/gkac1052

- [BioServices](https://doi.org/10.1093/bioinformatics/btt547)

  > Cokelaer, T, Pultz, D, Harder, LM, Serra-Musach, J, Saez-Rodriguez, J. (2013). BioServices: a common Python package to access biological Web Services programmatically. https://doi.org/10.1093/bioinformatics/btt547

- [pandas](https://zenodo.org/record/8239932)

  >  Data structures for statistical computing in python. https://doi.org/10.25080/Majora-92bf1922-00a

- [OrthoFinder](https://doi.org/10.1186/s13059-015-0721-2)

  > Emms DM, Kelly S. (2015). OrthoFinder: solving fundamental biases in whole genome comparisons dramatically improves orthogroup inference accuracy. https://doi.org/10.1186/s13059-015-0721-2

- [OrthoFinder2](https://doi.org/10.1186/s13059-019-1832-y)

  > Emms DM, Kelly S. (2019). OrthoFinder: phylogenetic orthology inference for comparative genomics. https://doi.org/10.1186/s13059-019-1832-y

- [DIAMOND](https://doi.org/10.1038/s41592-021-01101-x)

  > Buchfink B, Reuter K, Drost HG. (2021). Sensitive protein alignments at tree-of-life scale using DIAMOND. https://doi.org/10.1038/s41592-021-01101-x

- [cogeqc](https://github.com/almeidasilvaf/cogeqc)

  > Almeida-Silva F, Van de Peer Y. (2022). cogeqc: Systematic quality checks on comparative genomics analyses. R package version 1.3.0. https://github.com/almeidasilvaf/cogeqc

- [InterPro](https://doi.org/10.1093/nar/gkac993)

  > Paysan-Lafosse T, Blum M, Chuguransky S, Grego T, Pinto BL, Salazar GA, Bileschi ML, Bork P, Bridge A, Colwell L, Gough, J. (2023). InterPro in 2022. https://doi.org/10.1093/nar/gkac993

- [MAFFT](https://doi.org/10.1093/bioinformatics/bty121)

  > Nakamura, T, Yamada, KD, Tomii, K, & Katoh, K. (2018). Parallelization of MAFFT for large-scale multiple sequence alignments. https://doi.org/10.1093/bioinformatics/bty121

- [WITCH](https://doi.org/10.1089/cmb.2021.0585)

  > Shen, C, Park, M, & Warnow, T. (2022). WITCH: improved multiple sequence alignment through weighted consensus hidden Markov model alignment. https://doi.org/10.1089/cmb.2021.0585

- [ClipKIT](https://doi.org/10.1371/journal.pbio.3001007)

  > Steenwyk JL, Buida III TJ, Li Y, Shen XX, Rokas A. (2020). ClipKIT: a multiple sequence alignment trimming software for accurate phylogenomic inference. https://doi.org/10.1371/journal.pbio.3001007
  
- [CIALIGN](https://doi.org/10.7717/peerj.12983)

  > Tumescheit C, Firth AE, Brown K. (2022). CIAlign: A highly customisable command line tool to clean, interpret and visualise multiple sequence alignments. https://doi.org/10.7717/peerj.12983

- [FASTTREE2](https://doi.org/10.1371/journal.pone.0009490)

  > Price, MN, Dehal, SP, Arkin, AP. (2010). FastTree 2 – Approximately Maximum-Likelihood Trees for Large Alignments. https://doi.org/10.1371/journal.pone.0009490

- [IQ-TREE](https://doi.org/10.1093/molbev/msaa015)

  > Minh, BQ, Schmidt, HA, Chernomor, O, Schrempf, D, Woodhams, MD, Von Haeseler, A, Lanfear, R. (2020). IQ-TREE 2: new models and efficient methods for phylogenetic inference in the genomic era. https://doi.org/10.1093/molbev/msaa015

- [Asteroid](https://doi.org/10.1093/bioinformatics/btac832)

  > Morel B, Williams, TA, Stamatakis, A. (2023). Asteroid: a new algorithm to infer species trees from gene trees under high proportions of missing data. https://doi.org/10.1093/bioinformatics/btac832

- [GeneRax](https://doi.org/10.1093/molbev/msaa141)

  > Morel B, Kozlov AM, Stamatakis A, Szöllősi GJ. (2020). GeneRax: a tool for species-tree-aware maximum likelihood-based gene family tree inference under gene duplication, transfer, and loss. https://doi.org/10.1093/molbev/msaa141

- [SpeciesRax](https://doi.org/10.1093/molbev/msab365)

  > Morel B, Schade P, Lutteropp S, Williams TA, Szöllősi GJ, Stamatakis A. (2022). SpeciesRax: a tool for maximum likelihood species tree inference from gene family trees under duplication, transfer, and loss. https://doi.org/10.1093/molbev/msab365

- [tidyverse](https://doi.org/10.21105/joss.01686)

  > Wickham H et al. (2019). Welcome to the Tidyverse. https://doi.org/10.21105/joss.01686

- [reshape](https://doi.org/10.18637/jss.v021.i12)

  > Wickham H. (2007). Reshaping data with the reshape package. https://doi.org/10.18637/jss.v021.i12

- [cowplot](https://CRAN.R-project.org/package=cowplot)

  > Wilke C. (2020). cowplot: Streamlined plot theme and plot annotations for 'ggplot2'. R package version 1.1.1. https://CRAN.R-project.org/package=cowplot

- [elbow](https://github.com/ahasverus/elbow)
  > Casajus N. (2022). elbow: Detect inflection point of a concave curve. R package version 0.0.0.9000. https://github.com/ahasverus/elbow

- [ape](https://doi.org/10.1093/bioinformatics/btg412)
  > Paradis, E, Claude, J, Strimmer, K. (2004). APE: Analyses of Phylogenetics and Evolution in R language. https://doi.org/10.1093/bioinformatics/btg412

- [phytools](https://doi.org/10.1111/j.2041-210X.2011.00169.x)
  > Revell, LJ. (2011). phytools: an R package for phylogenetic comparative biology (and other things). https://doi.org/10.1111/j.2041-210X.2011.00169.x

- [DISCO](https://doi.org/10.1093/sysbio/syab070)
  > Willson, J, Saha Roddur, M, Liu, B, Zaharias, P, Warnow, T. (2021). DISCO: Species Tree Inference using Multicopy Gene Family Tree Decomposition. https://doi.org/10.1093/sysbio/syab070

## Software packaging/containerization tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. (2016). Computer software. Vers. 2-2.4.0. https://anaconda.com

- [Bioconda](https://doi.org/10.1038/s41592-018-0046-7)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. (2018). Bioconda: sustainable and comprehensive software distribution for the life sciences. https://doi.org/10.1038/s41592-018-0046-7

- [BioContainers](https://doi.org/10.1093/bioinformatics/btx192)

  > da Veiga Leprevost F. et. al. (2017). BioContainers: an open-source and community-driven framework for software standardization. https://doi.org/10.1093/bioinformatics/btx192

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

- [R](https://www.R-project.org/)

  > R Core Team. (2022). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria.

- [Singularity](https://doi.org/10.1371/journal.pone.0177459)
  
