# Arcadia-Science/phylorthology: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools - Software and Packages

- [BUSCO](https://doi.org/10.1002/cpz1.323)
  > Manni M, Berkeley MR, Seppey M, Zdobnov EM (2021). BUSCO: Assessing genomic data quality and beyond. Current Protocols, 1, e323. doi: 10.1002/cpz1.323

- [UniProt](https://doi.org/10.1093/nar/gkac1052)
  > The UniProt Consortium (2023). UniProt: the Universal Protein Knowledgebase in 2023

- [UniProt.ws](https://doi.org/doi:10.18129/B9.bioc.UniProt.ws)
  > Carlson M, Ramos M (2022). UniProt.ws: R Interface to UniProt Web Services. R package version 2.38.1.

- [OrthoFinder](https://doi.org/10.1186/s13059-015-0721-2)
  > Emms DM, Kelly S (2015). OrthoFinder: solving fundamental biases in whole genome comparisons dramatically improves orthogroup inference accuracy. Genome Biology 16:157

- [OrthoFinder2](https://doi.org/10.1186/s13059-019-1832-y)
  > Emms DM, Kelly S (2019). OrthoFinder: phylogenetic orthology inference for comparative genomics. Genome Biology 20:238

- [DIAMOND](https://doi.org/10.1038/s41592-021-01101-x)
  > Buchfink B, Reuter K, Drost HG (2021). Sensitive protein alignments at tree-of-life scale using DIAMOND. Nature Methods 18, 366–368. doi:10.1038/s41592-021-01101-x

- [cogeqc](https://github.com/almeidasilvaf/cogeqc)
  > Almeida-Silva F, Van de Peer Y (2022). cogeqc: Systematic quality checks on comparative genomics analyses. R package version 1.3.0.

- [InterPro](https://doi.org/10.1093/nar/gkac993)
  > Paysan-Lafosse T, Blum M, Chuguransky S, Grego T, Pinto BL, Salazar GA, Bileschi ML, Bork P, Bridge A, Colwell L, Gough, J (2023). InterPro in 2022. Nucleic Acids Research, 51(D1), pp.D418-D427.

- [UniProt](https://doi.org/10.1093/nar/gkac1052)
  > The UniProt Consortium (2023). UniProt: the Universal Protein Knowledgebase in 2023

- [MAFFT](https://doi.org/10.1093/bioinformatics/bty121)
  > Nakamura, T, Yamada, KD, Tomii, K, & Katoh, K (2018). Parallelization of MAFFT for large-scale multiple sequence alignments. Bioinformatics, 34(14), 2490-2492.

- [ClipKIT](https://doi.org/10.1371/journal.pbio.3001007)
  > Steenwyk, J. L., Buida III, T. J., Li, Y., Shen, X. X., & Rokas, A. (2020). ClipKIT: a multiple sequence alignment trimming software for accurate phylogenomic inference. PLoS biology, 18(12), e3001007.

- [IQ-TREE](https://doi.org/10.1093/molbev/msaa015)
  > Minh, BQ, Schmidt, HA, Chernomor, O, Schrempf, D, Woodhams, MD, Von Haeseler, A, Lanfear, R (2020). IQ-TREE 2: new models and efficient methods for phylogenetic inference in the genomic era. Molecular biology and evolution, 37(5), 1530-1534.

- [Asteroid](https://doi.org/10.1093/bioinformatics/btac832)
  > Morel B, Williams, TA, Stamatakis, A (2023). Asteroid: a new algorithm to infer species trees from gene trees under high proportions of missing data. Bioinformatics, 39(1), btac832.

- [GeneRax](https://doi.org/10.1093/molbev/msaa141)
  > Morel B, Kozlov AM, Stamatakis A, Szöllősi GJ (2020). GeneRax: a tool for species-tree-aware maximum likelihood-based gene family tree inference under gene duplication, transfer, and loss. Molecular biology and evolution, 37(9), 2763-2774.

- [SpeciesRax](https://doi.org/10.1093/molbev/msab365)
  > Morel B, Schade P, Lutteropp S, Williams TA, Szöllősi GJ, Stamatakis A (2022). SpeciesRax: a tool for maximum likelihood species tree inference from gene family trees under duplication, transfer, and loss. Molecular biology and evolution, 39(2), msab365.

- [tidyverse](https://doi.org/10.21105/joss.01686)
  > Wickham H, Averick M, Bryan J, Chang W, McGowan LDA, François R, ... & Yutani H (2019). Welcome to the Tidyverse. Journal of open source software, 4(43), 1686.

- [reshape](https://doi.org/10.18637/jss.v021.i12)
  > Wickham H (2007). Reshaping data with the reshape package. Journal of Statistical Software, 21(12).

- [cowplot](https://CRAN.R-project.org/package=cowplot)
  > Wilke C (2020). cowplot: Streamlined plot theme and plot annotations for 'ggplot2'. R package version 1.1.1.

- [elbow](https://github.com/ahasverus/elbow)
  > Casajus N (2022). elbow: Detect inflection point of a concave curve. R package version 0.0.0.9000.

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
